### PR TITLE
Add yaml-jsonpath implementation

### PR DESCRIPTION
**Pull request by glyn**
This PR was created as follows:
1. Added an implementation
2. Re-generated `build.ninja`
3. Ran `./src/wrap_in_docker.sh ninja`
4. Checked in all the new and changed files.

### CODE REVIEW

**Comment by cburgmer** [comment]
Ninja is running multiple processes in the background. When one fails, it lets the others finish before bailing. That means the failing one is unlikely to be the last one in the output.
Watch out for a big "FAILED" somewhere further up.

**Comment by glyn** [comment]
Thanks @cburgmer! The only "FAILED" is can see is in Proposal_A. Is that intended to be ok at HEAD?

**Comment by cburgmer** [comment]
I wonder. I don't regularly run the Docker version (too slow on a Mac), but it was on my list to check again. The Proposal_A is new, albeit very similar to the other JavaScript solutions. I'll try to reproduce tonight. It might be a trivial thing, so do have a look if you don't want to wait for my response.

**Comment by glyn** [comment]
@cburgmer Ok. My javascript knowledge is minimal, but I'll take a quick look for anything obvious.

**Comment by glyn** [comment]
Since I can no longer reproduce the failure in Proposal_A, here it is for posterity:
```
[33/7944] (cd proposals/Proposal_A && npm install) && echo > proposals/Proposal_A/build/node_modules_installed

> ohm-js@0.14.0 postinstall /json-path-comparison/proposals/Proposal_A/node_modules/ohm-js
> node bin/dev-setup.js

npm WARN jsonpath-proposal@1.0.0 No repository field.

added 32 packages from 77 contributors and audited 33 packages in 21.546s
found 1 low severity vulnerability
  run `npm audit fix` to fix them, or `npm audit` for details
[34/7944] LANG=en_US.UTF-8 LC_ALL= LC_COLLATE=C bash -c "diff --ignore-space-change <(echo 42) <(proposals/Proposal_A/run.sh ''\\$''[1]'' <<< '[0, 42]' | tr -d ' \n') || diff --ignore-space-change <(echo '[42]') <(proposals/Proposal_A/run.sh ''\\$''[1]'' <<< '[0, 42]' | tr -d ' \n')" > build/test_compilation/Proposal_A
FAILED: build/test_compilation/Proposal_A
LANG=en_US.UTF-8 LC_ALL= LC_COLLATE=C bash -c "diff --ignore-space-change <(echo 42) <(proposals/Proposal_A/run.sh ''\\$''[1]'' <<< '[0, 42]' | tr -d ' \n') || diff --ignore-space-change <(echo '[42]') <(proposals/Proposal_A/run.sh ''\\$''[1]'' <<< '[0, 42]' | tr -d ' \n')" > build/test_compilation/Proposal_A
internal/modules/cjs/loader.js:638
    throw err;
    ^

Error: Cannot find module './selector'
    at Function.Module._resolveFilename (internal/modules/cjs/loader.js:636:15)
    at Function.Module._load (internal/modules/cjs/loader.js:562:25)
    at Module.require (internal/modules/cjs/loader.js:692:17)
    at require (internal/modules/cjs/helpers.js:25:18)
    at Object.<anonymous> (/json-path-comparison/proposals/Proposal_A/index.js:2:19)
    at Module._compile (internal/modules/cjs/loader.js:778:30)
    at Object.Module._extensions..js (internal/modules/cjs/loader.js:789:10)
    at Module.load (internal/modules/cjs/loader.js:653:32)
    at tryModuleLoad (internal/modules/cjs/loader.js:593:12)
    at Function.Module._load (internal/modules/cjs/loader.js:585:3)
internal/modules/cjs/loader.js:638
    throw err;
    ^

Error: Cannot find module './selector'
    at Function.Module._resolveFilename (internal/modules/cjs/loader.js:636:15)
    at Function.Module._load (internal/modules/cjs/loader.js:562:25)
    at Module.require (internal/modules/cjs/loader.js:692:17)
    at require (internal/modules/cjs/helpers.js:25:18)
    at Object.<anonymous> (/json-path-comparison/proposals/Proposal_A/index.js:2:19)
    at Module._compile (internal/modules/cjs/loader.js:778:30)
    at Object.Module._extensions..js (internal/modules/cjs/loader.js:789:10)
    at Module.load (internal/modules/cjs/loader.js:653:32)
    at tryModuleLoad (internal/modules/cjs/loader.js:593:12)
    at Function.Module._load (internal/modules/cjs/loader.js:585:3)
```
There was also a build failure with Raku:
```
[15/7922] LANG=en_US.UTF-8 LC_ALL= LC_COLLATE=C bash -c "diff --ignore-space-change <(echo 42) <(implementations/Raku_JSON-Path/run.sh ''\\$''[1]'' <<< '[0, 42]' | tr -d ' \n') || diff --ignore-space-change <(echo '[42]') <(implementations/Raku_JSON-Path/run.sh ''\\$''[1]'' <<< '[0, 42]' | tr -d ' \n')" > build/test_compilation/Raku_JSON-Path
FAILED: build/test_compilation/Raku_JSON-Path
LANG=en_US.UTF-8 LC_ALL= LC_COLLATE=C bash -c "diff --ignore-space-change <(echo 42) <(implementations/Raku_JSON-Path/run.sh ''\\$''[1]'' <<< '[0, 42]' | tr -d ' \n') || diff --ignore-space-change <(echo '[42]') <(implementations/Raku_JSON-Path/run.sh ''\\$''[1]'' <<< '[0, 42]' | tr -d ' \n')" > build/test_compilation/Raku_JSON-Path
===SORRY!=== Error while compiling /json-path-comparison/implementations/Raku_JSON-Path/main.p6
Could not find JSON::Path in:
    inst#/root/.raku
    inst#/root/rakudo/share/perl6/site
    inst#/root/rakudo/share/perl6/vendor
    inst#/root/rakudo/share/perl6/core
    ap#
    nqp#
    perl5#
at /json-path-comparison/implementations/Raku_JSON-Path/main.p6:1
===SORRY!=== Error while compiling /json-path-comparison/implementations/Raku_JSON-Path/main.p6
Could not find JSON::Path in:
    inst#/root/.raku
    inst#/root/rakudo/share/perl6/site
    inst#/root/rakudo/share/perl6/vendor
    inst#/root/rakudo/share/perl6/core
    ap#
    nqp#
    perl5#
at /json-path-comparison/implementations/Raku_JSON-Path/main.p6:1
```
Finally, there was an error in my implementation, but since it runs fine locally, I'll have to dig deeper.

**Comment by glyn** [comment]
I managed to reproduce the Raku failure locally, but I couldn't find `zef` anywhere quickly, so I gave up and hope that problem reproduces for you.

I don't yet know what the problem is with my implementation as this works both locally and inside an ubuntu container, but I can dig into this further.

**Comment by glyn** [comment]
The Proposal_A problem came back, so it seems to be intermittent.

**Comment by cburgmer** [comment]
I think you might be seeing a few things:
- Ninja fails as quickly as possible, finishing remaining threads
- If re-run, it will sometimes continue on other tasks successfully. In my experience this seems to happen related to tasks that have all preceding dependencies resolved from the previous run. I guess Ninja will sort tasks to bubble up those it can resolve without dependencies. It then looks as if Ninja has recovered, but will fail later once it comes back to tackle the previously failing task.

I might be wrong on the previous, but I'm sure on the following points:

The Proposal_A failure stems from `flatMap` being used, which has only been introduced after the version that's currently supported in Ubuntu (Node 11 ships it https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/flatMap). I've started a PR where that's hopefully fixed: https://github.com/cburgmer/json-path-comparison/pull/39

The Raku/Zef thing seems to a different issue: Zef on the first run will happily (albeit slowly) install all dependencies globally. When ninja fails (or generally terminates) the Docker container will be removed, and any changes done in the meantime lost.
On a re-run now, Ninja thinks it has already solved the Zef dependencies and will move on, then failing because the dependencies are missing.
A workaround for now is to do a `rm implementations/Raku_JSON-Path/build/zef_installed` everytime before re-running Ninja.

I'll raise another issue for this: https://github.com/cburgmer/json-path-comparison/issues/40
I'll describe a bit more of the reasoning there.

Feel free to wait for me to fix those issues, but if you pull in my NodeJS fix, and follow the Zef workaround you might be good to move on.

**Comment by cburgmer** [comment]
#39 fixes the Proposal_A and Raku issue. ninja finishes successfully.
I can merge those fixes into master. Let me know if that works for you.

**Comment by glyn** [comment]
@cburgmer After https://github.com/cburgmer/json-path-comparison/pull/39 lands and I've rebased, this PR will be ready for review. I'm happy to take any comments in the meantime, particularly on my [method](https://github.com/cburgmer/json-path-comparison/pull/38#issue-423865420).

File: bug_reports/Golang_github.com-vmware-labs-yaml-jsonpath.md [hunk 1]
```diff
@@ -0,0 +1,7 @@
+Results do not match other implementations
+
+The following queries provide results that do not match those of other implementations of JSONPath
+(compare https://cburgmer.github.io/json-path-comparison/):
```
Comment (glyn) [inline bug_reports/Golang_github.com-vmware-labs-yaml-jsonpath.md:4]: @cburgmer Not sure what this file is telling me. Could it be that my implementation doesn't match for a selector of the empty string??

File: bug_reports/Bash_JSONPath.sh.md [hunk 2]
```diff
@@ -61,7 +61,7 @@ The following queries provide results that do not match those of other implement
   ```
   Error:
   ```
-  ggrep: Invalid range end
+  grep: Invalid range end
```
Comment (cburgmer) [inline bug_reports/Bash_JSONPath.sh.md:64]: You'll see some change like that, due to Linux and OSX being slightly different. Safe to ignore/commit.

Nobody's been bothered enough so far to address this :)

File: bug_reports/Golang_github.com-vmware-labs-yaml-jsonpath.md (hunk 1 above)
Comment (cburgmer) [reply bug_reports/Golang_github.com-vmware-labs-yaml-jsonpath.md:4]: If this file is empty apart from the boilerplate, then it looks like you implementation is fully aligned with the current consensus! That's great news!

Comment (glyn) [reply bug_reports/Golang_github.com-vmware-labs-yaml-jsonpath.md:4]: Thanks! It took a few days of tweaking to align with the consensus, so I'm pleased the consensus didn't change in the meantime.

File: implementations/Golang_github.com-vmware-labs-yaml-jsonpath/main.go [hunk 3]
```diff
@@ -0,0 +1,37 @@
...
+	}
+
+	jsonResult, errType, err := wrapper.Find(selector, data)
+	if errType != wrapper.Ok {
+		fmt.Fprintln(os.Stderr, err)
+		switch errType {
+		case wrapper.NotSupported:
+			os.Exit(2) // not supported
+		case wrapper.NotFound:
+			os.Exit(1) // not found
```
Comment (cburgmer) [inline implementations/Golang_github.com-vmware-labs-yaml-jsonpath/main.go:30]: Got it the wrong way I think?

3 is not found (https://github.com/cburgmer/json-path-comparison/blob/master/src/query_implementation.sh#L64), 1 is general error.

**Comment by cburgmer** [comment]
One minor issue but otherwise looks great!

You haven't swayed any consensus yet as far as I can see, otherwise you would have seen a lot of changes in the other implementations' reports.

File: implementations/Golang_github.com-vmware-labs-yaml-jsonpath/main.go (hunk 3 above)
Comment (glyn) [reply implementations/Golang_github.com-vmware-labs-yaml-jsonpath/main.go:30]: Ah, my mistake. I'll fix that now.

**Comment by glyn** [comment]
@cburgmer Fix that issue. Please merge at will. Thanks for your help!

**Comment by cburgmer** [comment]
Thanks for being patient with the quirks in the Docker setup!
And great to see more well behaving JSONPath implementations :)

**Pull request merged**